lib/preload.js

```diff
@@ -352,32 +351,72 @@ class Preloader extends EventEmitter {
 
 	_prepareErrorHandler() {
 		// Emit an error event if the python script exits with an error
-		this.container.wait().then((data) => {
-			if (data.StatusCode !== 0 && data.StatusCode !== EXIT_CAUSED_BY_SIGINT) {
-				const output = this.bufferedStderr.getData().toString('utf8').trim();
-				let error;
-				if (
-					output.indexOf(GRAPHDRIVER_ERROR) !== -1 &&
-					output.indexOf(DOCKERD_USES_OVERLAY) !== -1
-				) {
-					error = new this.balena.errors.BalenaError(OVERLAY_MODULE_MESSAGE);
-				} else {
-					error = new Error(output);
-					// @ts-ignore
-					error.code = data.StatusCode;
+		this.container
+			.wait()
+			.then((data) => {
+				if (data.StatusCode !== 0) {
+					const output = this.bufferedStderr.getData().toString('utf8').trim();
+					let error;
+					if (
+						output.indexOf(GRAPHDRIVER_ERROR) !== -1 &&
+						output.indexOf(DOCKERD_USES_OVERLAY) !== -1
+					) {
+						error = new this.balena.errors.BalenaError(OVERLAY_MODULE_MESSAGE);
+					} else {
+						error = new Error(output);
+						// @ts-ignore
+						error.code = data.StatusCode;
+					}
+					this.emit('error', error);
 				}
-				this.emit('error', error);
-			}
-		});
+			})
+			.catch((error) => this.emit('error', error));
 	}
 
+	/**
+	 * Send a command to `preload.py` (running in a Docker container) through
+	 * its stdin, and read a response from its stdout. The stdout response
+	 * must be a JSON object containing a `StatusCode` field (number / integer)
+	 * and an `Error` field (string). If `StatusCode` is the number zero, the
+	 * command execution is considered successful, otherwise it is assumed to
+	 * have failed and the returned promise is rejected with an error message
+	 * including the message provided in the `Error` field (if any).
+	 *
+	 * @param {string} command
+	 * @param {string[]} parameters
+	 */
 	_runCommand(command, parameters) {
 		return new Bluebird((resolve, reject) => {
+			const cmd = JSON.stringify({ command, parameters }) + '\n';
 			this.stdout.once('error', reject);
 			this.stdout.once('data', (data) => {
-				resolve(JSON.parse(data).result);
```

Comment:
As I understood it... This `_runCommand` function is one "end" (the CLI process end) of the interprocess communication between:
* The balena CLI / `preload.js` process at one end, and
* `preload.py` running in a Docker container at the other end.

The interprocess communication uses JSON objects exchanged via stdin and stdout. However, in case `preload.py` failed for any reason, a JSON object might not be generated in its stdout, and this `JSON.parse(data).result` call could fail synchronously (rather than producing a rejected promise).

In addition, the interprocess communication has to concept of "commands" that are sent via stdin to be executed in a Docker container, producing the result in stdout. If the command produced a JSON object with StatusCode different than zero (i.e. a failed execution), at least some commands like `get_image_info` would ignore the error: `_runCommand` would resolve rather than reject, and the resolved object's `StatusCode` would also not be inspected anywhere else (like elsewhere in `preload.js`, or elsewhere in `balena-cli`).

I though it would make sense for `_runCommand` to inspect `StatusCode` and produce a rejected promise with an error if `StatusCode` was not zero. This is what my changes to `_runCommand` implement.